2020_RTSS_Shuai_Zhao/src/parallel_provider_consumer.rs

```diff
@@ -5,35 +5,112 @@
+use std::collections::{HashMap, VecDeque};
 
 use lib::graph_extension::*;
 use petgraph::graph::{Graph, NodeIndex};
 
+/// See the second paragraph of IV. A. Concurrent provider and consumer model for a detailed explanation.
 /// Algorithm 1: Step1 identifying capacity providers.
 /// capacity provider is a sub paths of the critical path
-/// See the second paragraph of IV. A. Concurrent provider and consumer model for a detailed explanation.
-#[allow(dead_code)] // TODO: remove
-pub fn get_providers(mut dag: Graph<NodeData, f32>) -> Vec<Vec<NodeIndex>> {
-    let mut critical_path: VecDeque<NodeIndex> = dag.get_critical_path().into();
+pub fn get_providers(
+    dag: &Graph<NodeData, f32>,
+    critical_path: Vec<NodeIndex>,
+) -> Vec<Vec<NodeIndex>> {
+    let mut deque_critical_path = VecDeque::from(critical_path);
     let mut providers: Vec<Vec<NodeIndex>> = Vec::new();
-    while !critical_path.is_empty() {
-        let mut provider = vec![critical_path.pop_front().unwrap()];
-        while !critical_path.is_empty() && dag.get_pre_nodes(critical_path[0]).unwrap().len() == 1 {
-            provider.push(critical_path.pop_front().unwrap());
+    while !deque_critical_path.is_empty() {
+        let mut provider = vec![deque_critical_path.pop_front().unwrap()];
+        while !deque_critical_path.is_empty()
+            && dag.get_pre_nodes(deque_critical_path[0]).unwrap().len() == 1
+        {
+            provider.push(deque_critical_path.pop_front().unwrap());
         }
         providers.push(provider);
     }
     providers
 }
 
-#[cfg(test)]
+/// Algorithm 1: Step2 identifying capacity consumers.
+/// Capacity consumers represent specific non-critical nodes.
+/// F_consumers is a consumer set that can be simultaneous executed capacity providers, and whose execution delays the start of the next capacity providers.
+pub fn get_f_consumers(
+    dag: &mut Graph<NodeData, f32>,
+    critical_path: Vec<NodeIndex>,
+) -> HashMap<Vec<NodeIndex>, Vec<NodeIndex>> {
+    let mut providers = get_providers(dag, critical_path.clone());
+    let mut f_consumers: HashMap<Vec<NodeIndex>, Vec<NodeIndex>> = HashMap::new();
+    let mut non_critical_nodes = dag.get_non_critical_nodes(critical_path).unwrap();
+    let mut current_provider = providers.remove(0);
+    while !providers.is_empty() {
+        let next_provider = providers.remove(0);
+        let mut f_consumer = Vec::new();
 
-mod tests {
-    use std::collections::HashMap;
+        for next_provider_node in &next_provider {
+            let anc_nodes = dag.get_anc_nodes(*next_provider_node).unwrap();
+
+            for anc_node in anc_nodes {
+                if non_critical_nodes.contains(&anc_node) {
+                    f_consumer.push(anc_node);
+                }
+            }
+        }
+        f_consumers.insert(current_provider, f_consumer.clone());
+        current_provider = next_provider;
+        non_critical_nodes.retain(|&node_index| !f_consumer.contains(&node_index));
+    }
```

Comment:
The following implementation seems to be good.

```
    let providers = get_providers(dag, critical_path.clone());
    let mut f_consumers: HashMap<Vec<NodeIndex>, Vec<NodeIndex>> = HashMap::new();
    let mut non_critical_nodes: HashSet<_> = dag
        .get_non_critical_nodes(critical_path)
        .unwrap()
        .into_iter()
        .collect();

    for p_i in 0..providers.len() - 1 {
        let mut f_consumer = Vec::new();
        for next_provider_node in providers[p_i + 1].iter() {
            let anc_nodes = dag.get_anc_nodes(*next_provider_node).unwrap();
            for anc_node in anc_nodes {
                if non_critical_nodes.remove(&anc_node) {
                    f_consumer.push(anc_node);
                }
            }
        }
        f_consumers.insert(providers[p_i].clone(), f_consumer.clone());
    }
```

Reply:
Code Streamlining